src/cliffordsynthesis/encoding/MultiGateEncoder.cpp

```diff
@@ -132,19 +144,73 @@ void MultiGateEncoder::assertTwoQubitGateOrderConstraints(
   }
 
   // gate variables of the current and the next time step
-  const auto& current = vars.gC[pos][ctrl][trgt];
-  const auto& gSNow   = vars.gS[pos];
-  const auto& gCNext  = vars.gC[pos + 1];
+  const auto& gSNow  = vars.gS[pos];
+  const auto& gSNext = vars.gS[pos + 1];
+  const auto& gCNext = vars.gC[pos + 1];
 
   // two identical CNOTs may not be applied in a row because they would cancel.
-  lb->assertFormula(LogicTerm::implies(current, !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gCNext[trgt][ctrl]));
 
   // if no gate is applied to both qubits, no CNOT on them can be applied in the
   // next time step.
-  const auto noneIndex     = gateToIndex(qc::OpType::None);
-  const auto noGate        = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // no gate on both qubits => no CNOT on them in the next time step.
+  constexpr auto noneIndex = gateToIndex(qc::OpType::None);
+  const auto     noGate    = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // similarly,
+  // H - X   c - H
+  //     | = |
+  // H - c   X - H
+  // i.e., Hadamards on both qubits => no CNOT on them in the next time step.
+  constexpr auto hIndex = gateToIndex(qc::OpType::H);
+  const auto     hh     = gSNow[hIndex][ctrl] && gSNow[hIndex][trgt];
+
   const auto noFurtherCnot = !gCNext[ctrl][trgt] && !gCNext[trgt][ctrl];
-  lb->assertFormula(LogicTerm::implies(noGate, noFurtherCnot));
+  lb->assertFormula(LogicTerm::implies(noGate || hh, noFurtherCnot));
+
+  // No Xs on both qubits after CNOT
+  constexpr auto xIndex = gateToIndex(qc::OpType::X);
+  const auto     xx     = gSNext[xIndex][ctrl] && gSNext[xIndex][trgt];
+  const auto cnotBefore = vars.gC[pos][ctrl][trgt] || vars.gC[pos][trgt][ctrl];
+  lb->assertFormula(LogicTerm::implies(cnotBefore, !xx));
+
+  constexpr auto zIndex = gateToIndex(qc::OpType::Z);
+  const auto     zz     = gSNext[zIndex][ctrl] && gSNext[zIndex][trgt];
+  lb->assertFormula(LogicTerm::implies(cnotBefore, !zz));
+
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gSNext[xIndex][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gSNext[xIndex][ctrl]));
+
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gSNext[zIndex][ctrl]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gSNext[zIndex][trgt]));
+
+  constexpr auto sIndex = gateToIndex(qc::OpType::S);
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gSNext[sIndex][ctrl]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gSNext[sIndex][trgt]));
+}
```

Comment:
This part is missing some comments explaining the symmetry breaker. For example, I am not 100% convinced, that the last implication is true or at least guaranteed to preserve depth-optimality.

Furthermore, I think it is missing rules for the Y gate and the Sdag gate (if the S rule really holds), which should be analogous to the above rules.

Last but not least, many of these constraints can actually be combined so that multiple implications can be reduced to a single one. (`LogicTerm::implies(vars.gC[pos][ctrl][trgt], X ... Y ... Z)`)

Reply:
Very good catch. The rules should be all correct. Any Z-rotation commutes with the control a CNOT. But this might indeed break depth-optimality. Only the symmetry where there is a gate after both the control and target are guaranteed to preserve depth-optimality because no gates can be parallel after the CNOTs. I removed all those symmetry breakers.

On the other hand I included symmetry breakers for the case when two Z or X gates are **before** the CNot, because these can also just be replaced by a single gate after the CNOT without changing the depth.

Reply:
Yeah, exactly. With two-qubit gates, one has to always consider both qubits either before or after the two-qubit gate.
Is there a general rule of the form:
```
-- [I | X | Y | Z | H] -- q0 --
-- [I | X | Y | Z | H] -- q1 --

==

-- [q0 | q1] -- [I | X | Y | Z | H] --
-- [q1 | q0] -- [I | X | Y | Z | H] --
```
?

I am not sure how large the set of single qubit gates can really be. Can it also include the S and Sdag gate? Is no such general statement possible at all? What is the most general two-qubit "normal form"?

Reply:
The S gate cannot be commuted through the target qubit of the CX gate. If that were the case then we could push all single-qubit gates to the end of the circuit. This can only be done with Paulis however. So the current rules should be the most general.